packages/roosterjs-editor-core/lib/corePlugins/utils/InlineEntityHandlers/inlineEntityOnPluginEvent.ts

```diff
@@ -0,0 +1,195 @@
+import { DELIMITER_AFTER, DELIMITER_BEFORE, DelimiterType } from './constants';
+import { isDelimiter } from './isDelimiter';
+import {
+    ChangeSource,
+    EntityOperation,
+    PluginEventType,
+    PositionType,
+} from 'roosterjs-editor-types';
+import {
+    commitEntity,
+    createRange,
+    getEntityFromElement,
+    getEntitySelector,
+    isBlockElement,
+    matchesSelector,
+    Position,
+    safeInstanceOf,
+} from 'roosterjs-editor-dom';
+import type { Entity, EntityOperationEvent, IEditor, PluginEvent } from 'roosterjs-editor-types';
+
+const ZERO_WIDTH_SPACE = '\u200B';
+
+let delimiterId = 0;
+
+export function inlineEntityOnPluginEvent(event: PluginEvent, editor: IEditor) {
+    switch (event.eventType) {
+        case PluginEventType.EntityOperation:
+            const { wrapper } = event.entity;
+            switch (event.operation) {
+                case EntityOperation.NewEntity:
+                    // If the new entity is Read Only and inline, add delimiters to it
+                    if (isReadOnly(event.entity)) {
+                        const element = wrapper as HTMLElement;
+
+                        // Do not add delimiters if the entity already have them, in scenarios such as Remove + Undo
+                        if (isBetweenDelimiter(element)) {
+                            return;
+                        }
+
+                        const elementBefore = insertNode(
+                            `${ZERO_WIDTH_SPACE}`,
+                            element,
+                            'beforebegin'
+                        );
+                        const elementAfter = insertNode(`${ZERO_WIDTH_SPACE}`, element, 'afterend');
+
+                        delimiterId++;
+
+                        const idBefore = delimiterId + DELIMITER_BEFORE;
+                        const idAfter = delimiterId + DELIMITER_AFTER;
+                        commitEntity(elementBefore, idBefore, false, idAfter);
+                        commitEntity(elementAfter, idAfter, false, idBefore);
+
+                        editor.triggerContentChangedEvent(ChangeSource.InsertEntity, elementBefore);
+                        editor.triggerContentChangedEvent(ChangeSource.InsertEntity, elementAfter);
+
+                        editor.runAsync(() => {
+                            editor.select(new Position(wrapper, PositionType.After));
+                        });
+                    }
+                    break;
+                case EntityOperation.RemoveFromStart:
+                case EntityOperation.RemoveFromEnd:
+                case EntityOperation.Overwrite:
```

Comment:
Similarly, this can be part of the code where we detect and trigger these events.